arch/xtensa/core/xtensa_intgen.py

```diff
@@ -0,0 +1,130 @@
+        for i in ints:
+            # FIXME: a little work could allow us to extract the
+            # handler pointer and argument as literals, saving a few
+            # instructions and avoiding the need to link in
+            # _sw_isr_table entirely.
+            cprint("if (mask & (1 << " + str(i) + ")) {")
+            cprint("struct _isr_table_entry *e = &_sw_isr_table[" + str(i) + "];")
+            cprint("e->isr(e->arg);")
+            cprint("return 1 << " + str(i) + ";")
+            cprint("}")
+    else:
+        half = int(len(ints)/2)
+
+        m = 0
+        for i in ints[0:half]:
+            m |= 1 << i
+        cprint("if (mask & " + ("0x%x" % (m)) + ") {")
+        cprint("// interrupts: " + str(ints[0:half]))
+        emit_int_handler(ints[0:half])
+        cprint("} else {")
+        cprint("// interrupts: " + str(ints[half:]))
+        emit_int_handler(ints[half:])
+        cprint("}")
+
+########################################################################
+
+# Annoyingly need to join lines and remove #-marked annotations.  Some
+# versions of the preprocessor (ahem, esp32 SDK) like to include
+# newlines in the output where the original expressions are expanded
+# from 100% single line macros.  Slurp it into a single string and
+# parse via whitespace.
+blob = ""
+for l in fileinput.input():
+    l = l if l.find("#") < 0 else l[0:l.find("#")]
+    blob += l.rstrip() + " "
+
+for match in re.finditer(r'__xtensa_int_level_magic__\s+(\d+)\s+(\d+)', blob):
+    irq = int(match.group(1))
+    lvl = int(match.group(2))
+
+    if not lvl in ints_by_lvl:
+        ints_by_lvl[lvl] = []
+
+    ints_by_lvl[lvl].append(irq)
+
+cprint("/*")
+cprint(" * THIS FILE WAS AUTOMATICALLY GENERATED.  DO NOT EDIT.")
+cprint(" *")
+cprint(" * Functions here are designed to produce efficient code to")
+cprint(" * search an Xtensa bitmask of interrupts, inspecting only those bits")
+cprint(" * declared to be associated with a given interrupt level.  Each")
+cprint(" * dispatcher will handle exactly one flagged interrupt, in numerical")
+cprint(" * order (low bits first) and will return a mask of that bit that can")
+cprint(" * then be cleared by the calling code.  Unrecognized bits for the")
+cprint(" * level will invoke an error handler.")
+cprint(" */")
+cprint("")
+
+# Re-include the core-isa header and be sure our definitions match, for sanity
+cprint("#include <xtensa/config/core-isa.h>")
+cprint("#include <sw_isr_table.h>")
+cprint("")
+for l in ints_by_lvl:
+    for i in ints_by_lvl[l]:
+        v = "XCHAL_INT" + str(i) + "_LEVEL"
+        cprint("#if !defined(" + v + ") || " + str(v) + "!= " + str(l))
+        cprint("#error core-isa.h interrupt level does not match dispatcher!")
+        cprint("#endif")
+cprint("")
+
+# Populate empty levels just for sanity.  The second-to-last interrupt
+# level (usually "debug") typically doesn't have any associated
+# vectors, but we don't have any way to know that a-prioi.
+max = 0
```

Comment:
Python has a max() builtin that works pretty much with everything: you can pass various arguments or any iterable object. So this loop can be written as `max_level = max(ints_by_lvl)`.